napari/_qt/_qplugins/_qnpe2.py

```diff
@@ -0,0 +1,210 @@
+        widget_callable
+    ):
+        widget_param = ""
+    else:
+        raise TypeError(
+            trans._(
+                "'{widget}' must be `QtWidgets.QWidget` or `magicgui.widgets.Widget` subclass, `MagicFactory` instance or function. Please raise an issue in napari GitHub with the plugin and widget you were trying to use.",
+                deferred=True,
+                widget=widget_name,
+            )
+        )
+    return widget_param
+
+
+def _toggle_or_get_widget(
+    napari_viewer: Viewer,
+    mf: PluginManifest,
+    widget_name: str,
+    name: str,
+) -> Optional[Tuple[Union[FunctionGui, QWidget, Widget], str]]:
+    """Toggle if widget already built otherwise return widget.
+
+    Returned widget will be added to main window by a processor.
+    Note for magicgui type widget contributions, `Viewer` injection is done by
+    `magicgui.register_type` instead of a provider via annnotation.
+    """
+    window = napari_viewer.window
+    if dock_widget := window._dock_widgets.get(name):
+        dock_widget.setVisible(not dock_widget.isVisible())
+        return None
+
+    # Get widget param name (if any) and check type
+    if widget_contribution := get_widget_contribution(
+        mf.name,
+        widget_name,
+    ):
+        widget_callable, _ = widget_contribution
+        widget_param = _get_widget_viewer_param(widget_callable, widget_name)
+
+        kwargs = {}
+        if widget_param:
+            kwargs[widget_param] = napari_viewer
+        return widget_callable(**kwargs), name
+    return None
+
+
+def _get_widgets_submenu_actions(
+    mf: PluginManifest,
+) -> Tuple[List[Tuple[str, SubmenuItem]], List[Action]]:
+    """Get widget submenu and actions for a single npe2 plugin manifest."""
+    # If no widgets, return
+    if not mf.contributions.widgets:
+        return [], []
+
+    widgets = mf.contributions.widgets
+    multiprovider = len(widgets) > 1
+    submenu_id, submenu = _get_contrib_parent_menu(
+        multiprovider,
+        MenuId.MENUBAR_PLUGINS,
+        mf,
+        MenuGroup.PLUGIN_MULTI_SUBMENU,
+    )
+
+    widget_actions: List[Action] = []
+    for widget in widgets:
+        full_name = menu_item_template.format(
+            mf.display_name,
+            widget.display_name,
+        )
+
+        def _widget_callback(
+            napari_viewer: Viewer,
+            widget_name: str = widget.display_name,
+            name: str = full_name,
+        ) -> Optional[Tuple[Union[FunctionGui, QWidget, Widget], str]]:
+            return _toggle_or_get_widget(
+                napari_viewer=napari_viewer,
+                mf=mf,
+                widget_name=widget_name,
+                name=name,
+            )
+
+        def _get_current_dock_status(
+            window: Window,
+            name: str = full_name,
+        ) -> bool:
+            if name in window._dock_widgets:
+                return window._dock_widgets[name].isVisible()
+            return False
```

Comment:
I can take it out and define a function that calls here, just like what we're doing in `_widget_callback`. I am not expert on memory leaking. How would this fix memory leaking?

Reply:
We have to define a `callback` to pass to `Action`. I think from a minimising references case, we would want to define this function in the module and use a lambda to pass the args, BUT I don't think we're meant to use lambdas. I forget exactly why, I'd have to dig into old PRs. @DragaDoncila may know.

Reply:
Okay nevermind, we avoid lambdas due to leaking. So the only option would be:

> I can take it out and define a function that calls here, just like what we're doing in `_widget_callback`

Would that reduce the leaking risk?

Reply:
In general, when you define an inline function in python, its context contains all variables from the function where it is defined. 

As often such functions are short living, it is not a problem. But here we have functions that are connected to signals. So they will keep alive all objects referred from parent function body. Even if now there will be a problematic object (storing huge memory), it could be introduced in the future. Because of this, we avoid connecting lambdas and local functions to signals.

Reply:
I cannot think of a good solution to this. We need to create a function for each widget.

> So they will keep alive all objects referred from parent function body

I assume strings would be fine? I could get the viewer/window inside the inner function, so it is not referred from parent function body?

Reply:
Have amended this to use partial, which works. THanks @DragaDoncila @jni